distributed/comm/core.py

```diff
@@ -0,0 +1,292 @@
+
+from tornado import gen
+from tornado.ioloop import IOLoop
+
+from ..metrics import time
+
+
+logger = logging.getLogger(__name__)
+
+# Connector instances
+
+connectors = {
+    #'tcp': ...,
+    #'zmq': ...,
+    }
+
+
+# Listener classes
+
+listeners = {
+    #'tcp': ...,
+    # 'zmq': ...,
+    }
+
+
+DEFAULT_SCHEME = 'tcp'
+
+
+class CommClosedError(IOError):
+    pass
+
+
+class Comm(with_metaclass(ABCMeta)):
+    """
+    A message-oriented communication object, representing an established
+    communication channel.  There should be only one reader and one
+    writer at a time: to manage current communications, even with a
+    single peer, you must create distinct ``Comm`` objects.
+
+    Messages are arbitrary Python objects.  Concrete implementations
+    of this class can implement different serialization mechanisms
+    depending on the underlying transport's characteristics.
+    """
+
+    # XXX add set_close_callback()?
+
+    @abstractmethod
+    def read(self, deserialize=None):
+        """
+        Read and return a message (a Python object).  If *deserialize*
+        is not None, it overrides this communication's default setting.
+
+        This method is a coroutine.
+        """
+
+    @abstractmethod
+    def write(self, msg):
+        """
+        Write a message (a Python object).
+
+        This method is a coroutine.
+        """
+
+    @abstractmethod
+    def close(self):
+        """
+        Close the communication cleanly.  This will attempt to flush
+        outgoing buffers before actually closing the underlying transport.
+
+        This method is a coroutine.
+        """
+
+    @abstractmethod
+    def abort(self):
+        """
+        Close the communication immediately and abruptly.
+        Useful in destructors or generators' ``finally`` blocks.
+        """
+
+    @abstractmethod
+    def closed(self):
+        """
+        Return whether the stream is closed.
+        """
+
+    @abstractproperty
+    def peer_address(self):
+        """
+        The peer's address.  For logging and debugging purposes only.
+        """
+
+
+class Listener(with_metaclass(ABCMeta)):
+
+    @abstractmethod
+    def start(self):
+        """
+        Start listening for incoming connections.
+        """
+
+    @abstractmethod
+    def stop(self):
+        """
+        Stop listening.  This does not shutdown already established
+        communications, but prevents accepting new ones.
+        """
+        tcp_server, self.tcp_server = self.tcp_server, None
+        if tcp_server is not None:
+            tcp_server.stop()
+
+    @abstractproperty
+    def listen_address(self):
+        """
+        The listening address as a URI string.
+        """
+
+    @abstractproperty
+    def contact_address(self):
+        """
+        An address this listener can be contacted on.  This can be
+        different from `listen_address` if the latter is some wildcard
+        address such as 'tcp://0.0.0.0:123'.
+        """
```

Comment:
It might be good to add these classes and their methods as API documentation within the docs.  They provide an informative explanation of what one needs to do to implement another transport mechanism.